### ISSUE TITLE
Maybe a terrible mistake

### ISSUE THREAD

**Issue by rockeyben** [experiencor/keras-yolo2#353]
```
conf_mask = conf_mask + tf.to_float(best_ious < 0.6) * (1 - y_true[..., 4]) * self.no_object_scale
conf_mask = conf_mask + y_true[..., 4] * self.object_scale
nb_conf_box = tf.reduce_sum(tf.to_float(conf_mask > 0.0))
loss_conf = tf.reduce_sum(tf.square(true_box_conf-pred_box_conf) * conf_mask) / (nb_conf_box + 1e-6) / 2.
```
If you do this in this way, rather than compute neg conf loss and pos conf loss seperately, the neg conf loss will overwhelm the pos conf loss, and the model will converge to a bad status.
Just a simple case:
```
3*x/x + 5*y/y != (3*x+5*y)/(x+y)
```
And if one of x or y is too large, the imbalance will be more significant

**Comment by rockeyben** [comment]
I can't use your model to fit 4 images, but if I seperate the neg conf loss and pos conf loss, it works.
Just FYI

**Comment by rodrigo2019** [comment]
Did you compared a trainning with your modifications vs the original? could you share the results?

**Comment by rockeyben** [comment]
@rodrigo2019 
Sorry, I am doing other work recently so I can't conclude my result until now. But I am doing experiments in a different setteing. I am trying to adjust this repo to implement bird eye view car detection on KITTI. 
[ECCV18: YOLO3D: End-to-end real-time 3D Oriented Object Bounding Box Detection from LiDAR Point Cloud](https://arxiv.org/abs/1808.02350)
In this task, we have to detect 3D box from bird eye view, which means the car is relatively small, and positive part  of output score map is very small. It should be mentioned that author adjust darknet from 1/32 downsampling to 1/16 downsampling to overcome this situation, but it still makes the imbalance between pos and neg part in scoremap even worse. (neg part overwhelm pos part)

So I use this repo's loss function and I find no matter how low the loss become, this network still can't perform well on only 4 images (performs well on 1 image, but fails on others). First I try to change the parameter of neg and pos loss, so that the imbalace will be overcomed, but I find it doesn't work even I change the ratio to 20:1(neg:pos). So I changed the loss function to following version:
@bdhammel 
``` python
conf_mask_neg = tf.to_float(best_ious < 0.6) * (1 - y_true[..., 4]) * self.no_object_scale
conf_mask_pos = y_true[..., 4] * self.object_scale
nb_conf_box_neg = tf.reduce_sum(tf.to_float(conf_mask_neg > 0.0))
nb_conf_box_pos = tf.reduce_sum(tf.to_float(conf_mask_pos > 0.0))
loss_conf_neg = tf.reduce_sum(tf.square(true_box_conf-pred_box_conf) * conf_mask_neg) / (nb_conf_box_neg + 1e-6) / 2.
loss_conf_pos = tf.reduce_sum(tf.square(true_box_conf-pred_box_conf) * conf_mask_pos) / (nb_conf_box_pos + 1e-6) / 2.
loss_conf = loss_conf_neg + loss_conf_pos
```
And I find it finally performs normally.(even under ratio of 5:3, neg:pos)

I haven't done experiments on PASCAL VOC or COCO, though in this setting, the output score map is only 13x13 (my version is 38x38), and target objects are often large in the image, so the problem may not be that serious, but I think the problem still exists. If you are doing experiments on VOC and COCO, welcome to post your result here, thanks @bdhammel @rodrigo2019 

Best

**Comment by bdhammel** [comment]
Hey Rockey, thanks for the quick reply. This is the same approach I tried but, unfortunately, the network still has issues overfitting to a small dataset. These were the only modifications you made?

**Comment by rodrigo2019** [comment]
thank you for sharing, I will try to test in this week.

**Comment by rodrigo2019** [comment]
Could you share the full code? I dont know exactly where I need to insert this code.

**Comment by rodrigo2019** [comment]
@bdhammel, you are saying that with your modifications you got a better map, but when you make sanity check ploting the dataset you saw worst results?

**Comment by rockeyben** [comment]
It's weird, could you check your visualization code? If mAP is OK, I think visualization should be normal too.

**Comment by rodrigo2019** [comment]
Did you checked the mAP evalatuation function? the IOU threshold is set to 0.3 instead 0.5

**Comment by rodrigo2019** [comment]
I made a training using these parameters, and using your modification in loss function
```json
"object_scale":         5.0 ,
"no_object_scale":      3.0,
"coord_scale":          1.0,
"class_scale":          1.0,
```

I got better mAP, from 0.875 to 0.901 in my own dataset (around 5000 objects from single class). But I got much higher false positive rate, and now I understand what @bdhammel  said, the results look like are not consistent with obtained mAP.

I will try to upload a example video later to show you guys.

**Comment by rockeyben** [comment]
Is it the problem of score threshold? Or you plotted the score of FP and find it is a 'real' FP with higher score than a TP

**Comment by bdhammel** [comment]
I'll start this with the disclaimer that I really have not had the chance to dive into this issue. This is a project I'm trying to juggle with my left-hand while l have a lot of other stuff going on, so I can't rule out the problem isn't due to "piloting error." And unfortunately, I won't be able to revisit it for a while to give it the correct attention. But, in case anyone wants to reproduce what I'm seeing:

- The visualization code [is the one provided](https://github.com/experiencor/keras-yolo2/blob/master/predict.py).
- Dataset is the [MERL Dataset](http://www.merl.com/demos/merl-shopping-dataset) with bounding boxes. 
- Train on a single image for 10 epochs, until loss=.56 mAP=1, category1 AP=1, category2 AP=1, and the [detections are significantly off](https://s3.amazonaws.com/merl-yolo-test-data/small_train/image1961_detected.jpg)
- [training image](https://s3.amazonaws.com/merl-yolo-test-data/small_train/image1961.jpg), [training label](https://s3.amazonaws.com/merl-yolo-test-data/small_trainLabel/image1961.xml)

**Comment by rockeyben** [comment]
@bdhammel 
I know what happens:
[Line 471 in frontend.py](https://github.com/experiencor/keras-yolo2/blob/4e8c85ce02435f136d4f4cfe930b4ccb759fbaf8/frontend.py#L471)
If we use
```python
yolo.predict()
```
We will enter function in [Line 77 in utils.py](https://github.com/experiencor/keras-yolo2/blob/4e8c85ce02435f136d4f4cfe930b4ccb759fbaf8/utils.py#L77)
```python
def decode_netout(netout, anchors, nb_class, obj_threshold=0.3, nms_threshold=0.3):
```
We can see if we send no parameter to this function, object threshold will be default 0.3. Actually through your visualization image, it seems that the true positive get score of 0.709, while other FP get around 0.3, so what you need to do is to turn up the object threshold to kill FP.

**Comment by rodrigo2019** [comment]
Should I use the model with best validation loss or best mAP? because the best mAP model looks doesnt care about the false positives.

ps: I edited the code to do a mAP validation for every epoch
